File: sdk/java/src/main/java/io/agentsafe/spl/Parser.java
package io.agentsafe.spl;

import java.util.ArrayList;
import java.util.List;

/**
 * Tokenizer and recursive-descent parser for SPL S-expressions.
 */
public final class Parser {
    private Parser() {}

    public static Node parse(String src) {
        List<String> tokens = tokenize(src.trim());
        if (tokens.isEmpty()) throw new SplException("unexpected EOF");
        int[] pos = {0};
        Node result = parseExpr(tokens, pos);
        if (pos[0] != tokens.size()) throw new SplException("extra tokens");
        return result;
    }

    private static Node parseExpr(List<String> tokens, int[] pos) {
        if (pos[0] >= tokens.size()) throw new SplException("unexpected EOF");
        String tok = tokens.get(pos[0]++);

        if (tok.equals("(")) {
            List<Node> items = new ArrayList<>();
            while (true) {
                if (pos[0] >= tokens.size()) throw new SplException("unterminated (");
                if (tokens.get(pos[0]).equals(")")) {
                    pos[0]++;
                    break;
                }
                items.add(parseExpr(tokens, pos));
            }
            return Node.list(items);
        }
        if (tok.equals(")")) throw new SplException("unexpected )");
        return parseAtom(tok);
    }

    private static Node parseAtom(String tok) {
        if (tok.equals("#t")) return Node.bool_(true);
        if (tok.equals("#f")) return Node.bool_(false);
        // Try number
        try {
            double d = Double.parseDouble(tok);
            return Node.num(d);
        } catch (NumberFormatException ignored) {}
        // Quoted string
        if (tok.startsWith("\"") && tok.endsWith("\"") && tok.length() >= 2) {
            return Node.str(tok.substring(1, tok.length() - 1));
        }
        // Symbol
        return Node.sym(tok);
    }

    static List<String> tokenize(String src) {
        List<String> tokens = new ArrayList<>();
        StringBuilder buf = new StringBuilder();
        boolean inStr = false;

        for (int i = 0; i < src.length(); i++) {
            char ch = src.charAt(i);
            if (inStr) {
                buf.append(ch);
                if (ch == '"') {
                    inStr = false;
                    tokens.add(buf.toString());
                    buf.setLength(0);
                }
                continue;
            }
            switch (ch) {
                case '"' -> {
                    flushBuf(buf, tokens);
                    buf.append('"');
                    inStr = true;
                }
                case '(', ')' -> {
                    flushBuf(buf, tokens);
                    tokens.add(String.valueOf(ch));
                }
                case ' ', '\n', '\t', '\r' -> flushBuf(buf, tokens);
                default -> buf.append(ch);
            }
        }
        flushBuf(buf, tokens);
        return tokens;
    }

    private static void flushBuf(StringBuilder buf, List<String> tokens) {
        String s = buf.toString().trim();
        if (!s.isEmpty()) {
            for (String word : s.split("\\s+")) {
                tokens.add(word);
            }
        }
        buf.setLength(0);
    }
}

File: sdk/csharp/AgentSafe.Spl/Parser.cs
namespace AgentSafe.Spl;

public static class Parser
{
    private const int MaxPolicyBytes = 65536; // 64 KB

    public static Node Parse(string src)
    {
        if (System.Text.Encoding.UTF8.GetByteCount(src) > MaxPolicyBytes)
            throw new SplException($"policy exceeds maximum size of {MaxPolicyBytes} bytes");
        var tokens = Tokenize(src.Trim());
        if (tokens.Count == 0) throw new SplException("unexpected EOF");
        var pos = new int[] { 0 };
        var result = ParseExpr(tokens, pos);
        if (pos[0] != tokens.Count) throw new SplException("extra tokens");
        return result;
    }

    private static Node ParseExpr(List<string> tokens, int[] pos)
    {
        if (pos[0] >= tokens.Count) throw new SplException("unexpected EOF");
        var tok = tokens[pos[0]++];

        if (tok == "(")
        {
            var items = new List<Node>();
            while (true)
            {
                if (pos[0] >= tokens.Count) throw new SplException("unterminated (");
                if (tokens[pos[0]] == ")")
                {
                    pos[0]++;
                    break;
                }
                items.Add(ParseExpr(tokens, pos));
            }
            return Node.List(items);
        }
        if (tok == ")") throw new SplException("unexpected )");
        return ParseAtom(tok);
    }

    private static Node ParseAtom(string tok)
    {
        if (tok == "#t") return Node.Bool(true);
        if (tok == "#f") return Node.Bool(false);
        if (double.TryParse(tok, System.Globalization.NumberStyles.Any,
            System.Globalization.CultureInfo.InvariantCulture, out var d))
            return Node.Num(d);
        if (tok.StartsWith('"') && tok.EndsWith('"') && tok.Length >= 2)
            return Node.Str(tok[1..^1]);
        return Node.Sym(tok);
    }

    private static List<string> Tokenize(string src)
    {
        var tokens = new List<string>();
        var buf = new System.Text.StringBuilder();
        bool inStr = false;

        foreach (var ch in src)
        {
            if (inStr)
            {
                buf.Append(ch);
                if (ch == '"')
                {
                    inStr = false;
                    tokens.Add(buf.ToString());
                    buf.Clear();
                }
                continue;
            }

            switch (ch)
            {
                case '"':
                    FlushBuf(buf, tokens);
                    buf.Append('"');
                    inStr = true;
                    break;
                case '(' or ')':
                    FlushBuf(buf, tokens);
                    tokens.Add(ch.ToString());
                    break;
                case ' ' or '\n' or '\t' or '\r':
                    FlushBuf(buf, tokens);
                    break;
                default:
                    buf.Append(ch);
                    break;
            }
        }
        FlushBuf(buf, tokens);
        return tokens;
    }

    private static void FlushBuf(System.Text.StringBuilder buf, List<string> tokens)
    {
        var s = buf.ToString().Trim();
        if (s.Length > 0)
        {
            foreach (var word in s.Split(' ', StringSplitOptions.RemoveEmptyEntries))
                tokens.Add(word);
        }
        buf.Clear();
    }
}
